### DISCUSSION TITLE
Tendermint IPLD Schemas and Codecs

### DISCUSSION THREAD

**Discussion by i-norden** [tendermint/tendermint#6877] [Ideas]
Hello, this is a bit of a duplication of https://github.com/cosmos/cosmos-sdk/discussions/9505

I'm bringing the discussion here as the majority of the types we are introducing IPLD support for are a part of the Tendermint data structure.

The tentative IPLD schema definitions can be found here: https://github.com/ipld/ipld/pull/111

The ongoing Go codec implementations can be found here: https://github.com/vulcanize/go-codec-dagcosmos

A diagram of the Tendermint DAG (as I understand it) can be found in that repo: https://github.com/vulcanize/go-codec-dagcosmos/blob/main/tendermint_dag.png

The primary things I would like to solicit feedback on are to (1) ensure I have proper/comprehensive coverage of the types present in the Tendermint DAG and (2) open discussion on how best to handle the more "exotic" (from IPLD perspective) Merkle data structures in Tendermint- the HeaderFieldTree and the PartTree.

As it currently stands, the Tendermint types we are introducing schemas and codecs for are:

Header (tree of header fields)
Block (PartTree)
LightBlock
Params (HashedParams)
Proposal
Validator (SimpleValidator)
ValidatorTree
Evidence (union that can be either LightClientAttackEvidence or DuplicateVoteEvidence)
EvidenceTree
TxTree (Tx uses Raw codec, don't need new codec for it)
Result (ResponseDeliverTx)
ResultTree
Commit (CommitSig)
CommitTree

Support for most of these types is non-contentious from an IPLD perspective, but I want to make sure I have included everything we need here and have not included things we don't need.

Support for the Header (HeaderFieldTree) and Block (PartTree) is unprecedented and potentially contentious, since headers and blocks are indirectly content-hash referenced by a root hash of a Merkle tree containing the separate header fields or block parts (respectively) as the leaf values. From a conventional cotent-addressing approach these objects do not form singular IPLD nodes but rather complete sub-DAGs, and a number of issues present with this representation when decoding/encoding.

This problem has been expanded on extensively in this github issue: https://github.com/vulcanize/go-codec-dagcosmos/issues/20 so I won't repeat the details here, but would love some eyes on this issue! Thanks!

**Comment by cmwaters** [comment]
Hey, I've briefly scrolled through the issues/PR you linked but I'm still not 100% sure what the requirements are for which Tendermint data structures to support. Looking at the data structures you've mentioned, `Proposal` strikes me as interesting because it is ephemeral and just within consensus itself (i.e. you can't really query it). Or where you referring to the `ProposerAddress`

**Reply by i-norden** [reply]
Thanks for taking a look @cmwaters! In general, we want to support all hash-linked Tendermint data structures. So starting with the BlockID, any structure that can be hash-linked down to from there we want to include as part of the IPLD DAG.
